event/system.go

```diff
@@ -0,0 +1,153 @@
+// recommended to handle events in a separate goroutine to not block the
+// listener goroutine.
+func NewEventSystem(ctx context.Context, eventBuffer int, logger logrus.FieldLogger) *System {
+	return &System{
+		ctx:         ctx,
+		subscribers: make(map[Type]map[uint64]chan *Event),
+		eventBuffer: eventBuffer,
+		logger:      logger, // TODO: Log some debug events
+	}
+}
+
+// Subscribe to one or more events. It returns a subscriber ID that can be
+// used to unsubscribe, and an Event channel to receive events.
+// It panics if events is empty.
+func (s *System) Subscribe(events ...Type) (subID uint64, eventsCh <-chan *Event) {
+	if len(events) == 0 {
+		panic("must subscribe to at least 1 event type")
+	}
+	s.subMx.Lock()
+	defer s.subMx.Unlock()
+	s.subIDCount++
+	subID = s.subIDCount
+	evtCh := make(chan *Event, s.eventBuffer)
+	for _, evt := range events {
+		if s.subscribers[evt] == nil {
+			s.subscribers[evt] = make(map[uint64]chan *Event)
+		}
+		s.subscribers[evt][subID] = evtCh
+	}
+	return subID, evtCh
+}
+
+// Emit the event to all subscribers of its type.
+// It returns a function that can be optionally used to wait for all subscribers
+// to process the event (by signalling via the Done method).
+func (s *System) Emit(event *Event) (wait func(time.Duration) error) {
+	s.subMx.RLock()
+	defer s.subMx.RUnlock()
+	totalSubs := len(s.subscribers[event.Type])
+	if totalSubs == 0 {
+		return func(time.Duration) error { return nil }
+	}
+
+	if event.Done == nil {
+		event.Done = func() {}
+	}
+	origDoneFn := event.Done
+	doneCh := make(chan struct{}, s.eventBuffer)
+	doneFn := func() {
+		origDoneFn()
+		select {
+		case doneCh <- struct{}{}:
+		default:
+		}
+	}
+	event.Done = doneFn
+
+	for _, evtCh := range s.subscribers[event.Type] {
+		select {
+		case evtCh <- event:
+		default:
+		}
+	}
+
+	return func(timeout time.Duration) error {
+		var (
+			doneCount int
+			tout      = time.After(timeout)
+		)
+		for {
+			if doneCount == totalSubs {
+				close(doneCh)
+				return nil
+			}
+			select {
+			case <-doneCh:
+				doneCount++
+			case <-tout:
+				return fmt.Errorf("timed out after waiting %s for all '%s' events to be processed", timeout, event.Type)
+			case <-s.ctx.Done():
+				return nil
+			}
+		}
+	}
+}
+
+// Unsubscribe closes the Event channel and removes the subscription with ID
+// subID.
+func (s *System) Unsubscribe(subID uint64) {
+	s.subMx.Lock()
+	defer s.subMx.Unlock()
+	var seen bool
+	for _, sub := range s.subscribers {
+		if evtCh, ok := sub[subID]; ok {
+			if !seen {
+				close(evtCh)
+			}
+			delete(sub, subID)
+			seen = true
+		}
+	}
+}
+
+// UnsubscribeAll closes all event channels and removes all subscriptions.
+func (s *System) UnsubscribeAll() {
+	s.subMx.Lock()
+	defer s.subMx.Unlock()
+
+	seenSubs := make(map[uint64]struct{})
+	for _, sub := range s.subscribers {
+		for subID, evtCh := range sub {
+			if _, ok := seenSubs[subID]; !ok {
+				close(evtCh)
```

Comment:
But we do. That's the point of the `seenSubs` map and this check. We use a map instead of a bool here because we're dealing with more than one subscriber ID.

Reply:
Right :facepalm: